### PR TITLE
Feature/2017 propertydef migration

### LINKED ISSUE #2017: Migration of propetydef_v1
## Feature
Add migration of propertydef_v1 collection's entities.

### PR DESCRIPTION
**Pull request by dk1844**
~Property definitions are now part of by-dataset-names migration.~
 - ~entities are locked after successful migration~
 - ~non-locked property definitions are being found in datasets by name (even from not-locked datasets)~

 * Property definitions now have a separate mode of migration `-p`
 * DS migration contains a check that PDs of source must be found on target
   * currently not necessarily in the exact version - do we need to check target version to be == or >= to source PD version?

## Example run:

1. Initialize blank target db `mns2` first:
`./initialize_menas.py  mongodb://localhost:27017/admin -t mns2`
2. Migrate property definitions (here, using `-p *` for all PDs):
`./migrate_menas.py mongodb://localhost:27017/admin mongodb://localhost:27017/admin -s menas -p * -t mns2`
3. Now, datasets can be migrated:
`./migrate_menas.py mongodb://localhost:27017/admin mongodb://localhost:27017/admin -s menas -d mydataset1 mydataset2 -t mns2`

Merge after #2018 (based on it). just the last commit is relevant
Closes #2017

### CODE REVIEW

**Comment by dk1844** [comment]
- suggestions co-authored by @Zejnilovic 
   -  propdef-by-name migration (option `all`, too)
   - ds-by-name migration: optionally migrate propdefs by properties mentioned in DSs (default: false)

File: scripts/migration/menas_db.py [hunk 1]
```diff
@@ -146,6 +146,33 @@ def get_distinct_mapping_tables_from_ds_names(self, ds_names: List[str], migrati
...
+            }},
+            {"$unwind": "$props"},  # explodes each doc into multiple - each having prop key/val
+            {"$group": {
+                "_id": "notNeededButRequired",
+                "propkeys": {"$addToSet": "$props.k"}
+            }}
+        ])  # single doc with { _id: ... , "propkeys" : [propKey1, propKey2, ...]}
+
+        # if no props are present, the result may be empty
+        mapping_table_names_list = list(props_agg_result)  # cursor behaves one-iteration only.
```
Comment (dk1844) [inline scripts/migration/menas_db.py:169]: wrong var name (copy-pasta error)

Comment (dk1844) [reply scripts/migration/menas_db.py]: Fixed, thanks @Zejnilovic for pointing this out.

**Comment by benedeki** [comment]
We were discussing the possible strategies:

1. on dataset**s** migration start check that target has a superset of property definition of the source
2. on each dataset migration check if all properties defined have definition on the target system
   1. migrate the missing definitions
   2. skip the missing properties
   3. stop migration/skip the migration of the particular dataset

1 and 2.iii can prevent datasets migration
2.i can make some datasets on target system unrunable
2.ii means potential data-loss

2.i outcome can be mitigated by migrating such a missing property def as `optional`.

Also we should consider, that we do **control** the property definitions (only admins can define them) so discrepancies if once migrated should not really occur.

**Comment by dk1844** [comment]
selected #1 in wide-team collaboration sessions.

Implemented. **However, I am wondering if and if-so how the PD versions should be checked, too.**
 - Force PD versions to be equal? Or targetVersion to be >=  sourceVersion?

**Review by AdrianOlosutean — APPROVED** [review]
Code review

**Comment by Zejnilovic** [comment]
Funny thing and something to look out - if you have autocomplete and expansion then using `*` is dangerous 

```shell
╰─$ ./migrate_menas.py mongodb://localhost:27017 mongodb://localhost:27017 -s menas_e2e -p * -t migration_aws_5
Menas mongo migration
Running with settings: dryrun=False, verbose=False
Using migration #: '821bc8' and locking timestamp 2022-02-23 09:15:48.695849+00:00 (UTC)
  source connection-string: mongodb://localhost:27017
  source DB: menas_e2e
  target connection-string: mongodb://localhost:27017
  target DB: migration_aws_5
Checking source db validity...
Checking target db validity...
Property definition names supplied: ['__pycache__', 'constants.py', 'initialize_menas.py', 'menas_db.py', 'migrate_menas.py', 'requirements.txt']
Property definitions to migrate: []

No property definitions to migrate in propertydef_v1, skipping.
Done.
```

Property definitions supplied become all the files in the folder.

**Comment by Zejnilovic** [comment]
**Question:** I have 2 properties set/filled in on a Dataset. These properties' property definitions are not migrated. What should be the output of the error? Should it be a list of all the property definitions or just the two?

Stacktrace
```shell
Menas mongo migration
Running with settings: dryrun=False, verbose=False
Using migration #: '848aae' and locking timestamp 2022-02-23 09:37:03.280587+00:00 (UTC)
  source connection-string: mongodb://localhost:27017
  source DB: menas_e2e
  target connection-string: mongodb://localhost:27017
  target DB: migration_aws_5
Checking source db validity...
Checking target db validity...
Traceback (most recent call last):
  File "/some/path/enceladus/scripts/migration/./migrate_menas.py", line 367, in <module>
    run(args)
  File "/some/path/enceladus/scripts/migration/./migrate_menas.py", line 341, in run
    check_compatible_property_definitions(source_db, target_db)
  File "/some/path/enceladus/scripts/migration/./migrate_menas.py", line 304, in check_compatible_property_definitions
    raise MenasDbCollectionError(f"Property definitions missing on target: {diff},"
menas_db.MenasDbCollectionError: Property definitions missing on target: {'f', 'f2', 'firstProp3', 'firstProp123', 'TestEnumProperty2', 'firstProp123_1234_5678', 'firstProp2', 'firstProp123_1234_5678qwertyuiopasdfghjklzxcvbnmqwertyuiopasdfghjklzxcvb', 'TestEnumProperty', 'firstProp123_1234_5678qwertyuiopasdfghjklzxcvbnmqwertyuiopasdfghjklzxc', 'firstProp'}, cannot continue dataset migration! Migrate property definitions first (-p * or -p PDname1 PDname2 ...)
```

Even if I migrate the 2 needed property definitions I get the same error.

**Comment by dk1844** [comment]
> **Question:** I have 2 properties set/filled in on a Dataset. These properties' property definitions are not migrated. What should be the output of the error? Should it be a list of all the property definitions or just the two?
...

From the options outlined in https://github.com/AbsaOSS/enceladus/pull/2019#issuecomment-1032811921, #1 was implemented:
> 1. on dataset**s** migration start check that target has a superset of property definition of the source

This check is oblivious to what datasets are being migrated and what properties they actually use. The process is failed when any of the source property definitions are missing on target (by-name check), regardless if the selected datasets would specifically need these or not.

**Review by Zejnilovic — APPROVED** [review]

**Pull request merged**